### ISSUE TITLE
threshold experimentation

### ISSUE THREAD

**Issue by brandy333-umbre** [brandy333-umbre/medical-no-show-appointments#1]
The script scans thresholds from 0.0 → 1.0

Computes precision, recall, F1 at each

Selects:

Best threshold by F1

Threshold achieving recall ≥ 0.70

ROC-AUC → same

PR-AUC → same

F1 → slightly improved

Recall → improved

script:
from __future__ import annotations

import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    roc_auc_score,
    average_precision_score,
    confusion_matrix,
    classification_report,
    precision_score,
    recall_score,
    f1_score,
)

DATA_PATH = "data/noshow.csv"

def load_and_clean(path: str) -> pd.DataFrame:
    df = pd.read_csv(path).copy()

    # Target: "Yes" = no-show (1), "No" = showed up (0)
    df["target"] = (df["No-show"] == "Yes").astype(int)

    # Parse datetimes
    df["ScheduledDay"] = pd.to_datetime(df["ScheduledDay"], errors="coerce", utc=True)
    df["AppointmentDay"] = pd.to_datetime(df["AppointmentDay"], errors="coerce", utc=True)

    # Feature engineering
    df["lead_time_hours"] = (df["AppointmentDay"] - df["ScheduledDay"]).dt.total_seconds() / 3600.0
    df["scheduled_hour"] = df["ScheduledDay"].dt.hour
    df["scheduled_weekday"] = df["ScheduledDay"].dt.day_name()
    df["appointment_weekday"] = df["AppointmentDay"].dt.day_name()

    # Fix known data issue
    df = df[df["Age"].between(0, 120)]

    # Drop identifiers / leakage-prone columns + raw target/datetimes
    drop_cols = ["PatientId", "AppointmentID", "No-show", "ScheduledDay", "AppointmentDay"]
    df = df.drop(columns=[c for c in drop_cols if c in df.columns], errors="ignore")

    return df

def build_model(X: pd.DataFrame) -> Pipeline:
    # Column types (include "string" to avoid pandas warning in newer versions)
    num_cols = X.select_dtypes(include=["int64", "float64"]).columns.tolist()
    cat_cols = X.select_dtypes(include=["object", "string", "bool"]).columns.tolist()

    numeric_transformer = Pipeline(
        steps=[
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler()),
        ]
    )

    categorical_transformer = Pipeline(
        steps=[
            ("imputer", SimpleImputer(strategy="most_frequent")),
            ("onehot", OneHotEncoder(handle_unknown="ignore")),
        ]
    )

    preprocessor = ColumnTransformer(
        transformers=[
            ("num", numeric_transformer, num_cols),
            ("cat", categorical_transformer, cat_cols),
        ],
        remainder="drop",
    )

    clf = LogisticRegression(
        max_iter=2000,
        class_weight="balanced",
        solver="liblinear",
    )

    return Pipeline(
        steps=[
            ("preprocess", preprocessor),
            ("clf", clf),
        ]
    )

def evaluate_at_threshold(y_true: np.ndarray, proba: np.ndarray, threshold: float) -> dict:
    preds = (proba >= threshold).astype(int)
    return {
        "threshold": threshold,
        "precision": precision_score(y_true, preds, zero_division=0),
        "recall": recall_score(y_true, preds, zero_division=0),
        "f1": f1_score(y_true, preds, zero_division=0),
    }

def tune_thresholds(
    y_true: np.ndarray,
    proba: np.ndarray,
    target_recall: float = 0.70,
    n_thresholds: int = 101,
) -> tuple[pd.DataFrame, float, float]:
    thresholds = np.linspace(0.0, 1.0, n_thresholds)
    rows = [evaluate_at_threshold(y_true, proba, t) for t in thresholds]
    df = pd.DataFrame(rows)

    # Best threshold by F1
    best_f1_row = df.loc[df["f1"].idxmax()]
    best_f1_threshold = float(best_f1_row["threshold"])

    # Threshold achieving at least target recall with best precision (or closest if none)
    df_meet = df[df["recall"] >= target_recall]
    if len(df_meet) > 0:
        best_recall_threshold = float(df_meet.sort_values(["precision", "threshold"], ascending=[False, True]).iloc[0]["threshold"])
    else:
        # If target recall is impossible, choose the closest recall
        best_recall_threshold = float(df.iloc[(df["recall"] - target_recall).abs().idxmin()]["threshold"])

    return df, best_f1_threshold, best_recall_threshold

def print_eval(y_true: np.ndarray, proba: np.ndarray, threshold: float, title: str) -> None:
    preds = (proba >= threshold).astype(int)
    cm = confusion_matrix(y_true, preds)
    print(f"\n=== {title} ===")
    print(f"Threshold: {threshold:.2f}")
    print("Confusion matrix (TN FP / FN TP):")
    print(cm)
    print("\nClassification report:")
    print(classification_report(y_true, preds, digits=4))

def main() -> None:
    df = load_and_clean(DATA_PATH)
    X = df.drop(columns=["target"])
    y = df["target"].to_numpy()

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )

    model = build_model(X_train)
    model.fit(X_train, y_train)

    proba = model.predict_proba(X_test)[:, 1]

    roc_auc = roc_auc_score(y_test, proba)
    pr_auc = average_precision_score(y_test, proba)

    print("=== Evaluation (Threshold-independent) ===")
    print(f"ROC-AUC: {roc_auc:.4f}")
    print(f"PR-AUC:  {pr_auc:.4f}")

    # Tune thresholds
    metrics_df, best_f1_t, best_recall_t = tune_thresholds(y_test, proba, target_recall=0.70)

    # Show a quick summary of the best points
    best_f1_row = metrics_df.loc[metrics_df["f1"].idxmax()]
    print("\n=== Best threshold by F1 ===")
    print(best_f1_row.to_string(index=False))

    # Find the row used for target recall threshold (closest match)
    best_recall_row = metrics_df.iloc[(metrics_df["threshold"] - best_recall_t).abs().idxmin()]
    print("\n=== Threshold targeting recall >= 0.70 (best precision among those) ===")
    print(best_recall_row.to_string(index=False))

    # Print detailed confusion matrices / reports for both thresholds
    print_eval(y_test, proba, best_f1_t, "Evaluation at Best-F1 Threshold")
    print_eval(y_test, proba, best_recall_t, "Evaluation at Target-Recall Threshold (>= 0.70)")

    # Optional: save threshold table for inspection
    # metrics_df.to_csv("artifacts/threshold_metrics.csv", index=False)

if __name__ == "__main__":
    main()

output:
ROC-AUC: 0.6697
PR-AUC:  0.3067

=== Best threshold by F1 ===
0.470000
0.297806
0.665995
0.411573

=== Threshold targeting recall >= 0.70 (best precision among those) ===
0.450000
0.284779
0.721326
0.408344

=== Evaluation at Best-F1 Threshold ===
Threshold: 0.47
Confusion matrix (TN FP / FN TP):
[[10632  7010]
 [ 1491  2973]]

Classification report:
              precision    recall  f1-score   support

           0     0.8770    0.6027    0.7144     17642
           1     0.2978    0.6660    0.4116      4464

    accuracy                         0.6154     22106
   macro avg     0.5874    0.6343    0.5630     22106
weighted avg     0.7600    0.6154    0.6532     22106

=== Evaluation at Target-Recall Threshold (>= 0.70) ===
Threshold: 0.45
Confusion matrix (TN FP / FN TP):
[[9555 8087]
 [1244 3220]]

Classification report:
              precision    recall  f1-score   support

           0     0.8848    0.5416    0.6719     17642
           1     0.2848    0.7213    0.4083      4464

    accuracy                         0.5779     22106
   macro avg     0.5848    0.6315    0.5401     22106
weighted avg     0.7636    0.5779    0.6187     22106
